Fix tests

giuliano108 opened the pull request:
- fix both spec and acceptance tests
- add basic debugging output

Notes:

- Beaker doesn't work on the devboxes because `safer-docker` [gets in the way](https://github.com/puppetlabs/beaker/blob/3.8.0/lib/beaker/hypervisor/docker.rb#L110). As a workaround, I've added an "` && !ENV['DOCKER_HOST'].match(/safer_docker/)`" to the end of that line.
- In the output of `apt-get -s autoremove`, the code looks for lines starting with `Purg`. Unless the `--purge` CLI option or the `APT::Get::Purge` config option are used, dangling dependencies will be flagged with `Remv` instead. The code will never notice/purge them.

        root@de4b5b779e37:/# apt-get -s autoremove | grep dictd
          dict dictd dictzip libfile-copy-recursive-perl libmaa3 librecode0 m4 recode
        Remv dictd [1.12.1+dfsg-2]

    Any thoughts about what we should be doing here? Grep for either `Purg` or `Remv`? Tell the user that apt needs to be configured in a certain way for the module to work?

Tests output: https://i.fluffy.cc/FmGjt7rr41mddfQVR8X2RQ9KFNgkqTbM.html

Internal ticket: OPSIMP-632 _Test and roll out puppet dpkg holding strategy_.

/cc @solarkennedy, @oholiab

giuliano108 committed:
fix spec tests, port to RSpec 3 syntax

giuliano108 committed:
fix acceptance tests (just the framework side, tests still fail)

giuliano108 committed:
fix acceptance tests by enabling "purge by default"

solarkennedy commented:
Whoa!

solarkennedy commented:
I vote for `Purg` and have a class in this module that configures the apt preference for you, basically puppetize what you have in the test case:
```puppet
apt::conf { 'purge_all_the_things':
  content => '"APT::Get::Purge "true"',
}
```

Then add a it to the readme that you need to include it.

oholiab commented:
I agree with what @solarkennedy said!

Other than that, lgtm, shipit :)

giuliano108 committed:
manage APT::Get::Purge, better tests/comments

giuliano108 commented:
I've added a `package_purging::config` class even though I don't think `package_purging` should concern itself with configuring `apt.conf.d`. For example, our Puppet codebase sets `APT::Get::Purge` in the `apt_sources` module...
The acceptance tests are slightly better and I've fixed my previous comments that turned out to be incorrect.

If ShipIt, I'll bump the version we use at work to master (with due care and a internal CR), then move on to writing some tests for the `dpkg_hold` stuff (which is the real reason why I'm here). 

/cc @solarkennedy

solarkennedy commented:
I say ship!

solarkennedy commented:
@hashbrowncipher although I'm a contributor, I have yet to actually merge stuff into your repos on your behalf. Am I good to go to manage this repo and continue pushing the purging torch?

hashbrowncipher commented on lib/puppet/type/aptly_purge.rb:
```diff
@@ -52,7 +54,9 @@ def generate
     end
 
     managed_package_names = managed_packages.map(&:name)
+    Puppet.debug "managed_package_names: #{managed_package_names}"
```
nice!

hashbrowncipher commented:
Hey! Thanks for contributing. I'm going to change my mail filters to put Github PRs at the very top, so future delays shouldn't happen.

I also prefer just configuring apt.conf.d for you, over the other options.
